### PR TITLE
[Angular] Setting page titles natively with the Angular router

### LINKED ISSUE #19693: Angular 14 - Setting page titles natively with the Angular router
##### **Overview of the feature request**

Implements https://dev.to/brandontroberts/setting-page-titles-natively-with-the-angular-router-393j

##### **Related issues or PR**

- [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by qmonmert**
Fix #19693 

---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] The JDL part is updated if necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

File: generators/client/templates/angular/src/main/webapp/app/layouts/error/error.route.ts.ejs [hunk 1]
```diff
@@ -24,25 +24,17 @@ export const errorRoute: Routes = [
   {
     path: 'error',
     component: ErrorComponent,
-    data: {
-      pageTitle: 'error.title',
-    },
+    title: 'error.title',
```
Comment (Tcharl) [inline generators/client/templates/angular/src/main/webapp/app/layouts/error/error.route.ts.ejs:27]: What about being more explicit "title.key"

File: generators/client/templates/angular/src/main/webapp/app/template-page-title-strategy.ts.ejs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-%>
+import { Injectable } from "@angular/core";
+import { RouterStateSnapshot, TitleStrategy } from "@angular/router";
+import { TranslateService } from "@ngx-translate/core";
+
+@Injectable()
```
Comment (Tcharl) [inline generators/client/templates/angular/src/main/webapp/app/template-page-title-strategy.ts.ejs:23]: Shouldn't it be a directive instead?

Comment (qmonmert) [reply generators/client/templates/angular/src/main/webapp/app/template-page-title-strategy.ts.ejs:23]: no https://netbasal.com/handling-page-titles-in-angular-40b53823af4a

File: generators/client/templates/angular/src/main/webapp/app/layouts/error/error.route.ts.ejs (hunk 1 above)
Comment (qmonmert) [reply generators/client/templates/angular/src/main/webapp/app/layouts/error/error.route.ts.ejs:27]: I don't understand :/

Comment (Tcharl) [reply generators/client/templates/angular/src/main/webapp/app/layouts/error/error.route.ts.ejs:27]: Ah, sorry
```
{
    path: 'error',
    component: ErrorComponent,
    title: { key:...,
}}
```

**Comment by mshima** [comment]
Need to adjust the translation. I will try to commit the required change.
Captura de Tela 2022-09-19 às 18 50 42

File: generators/client/templates/angular/src/main/webapp/app/layouts/error/error.route.ts.ejs (hunk 1 above)
Comment (mshima) [reply generators/client/templates/angular/src/main/webapp/app/layouts/error/error.route.ts.ejs:27]: Would be interesting, but I don't think it's possible.
[https://angular.io/api/router/TitleStrategy `buildTitle](https://angular.io/api/router/TitleStrategy#buildtitle) returns a string.

File: generators/client/templates/angular/src/main/webapp/app/layouts/error/error.route.ts.ejs [hunk 3]
```diff
@@ -24,25 +24,17 @@ export const errorRoute: Routes = [
...
   },
   {
     path: 'accessdenied',
     component: ErrorComponent,
-    data: {
-      pageTitle: 'error.title',
-      errorMessage: 'error.http.403',
-    },
+    title: 'error.http.403',
```
Comment (mshima) [inline generators/client/templates/angular/src/main/webapp/app/layouts/error/error.route.ts.ejs:32]: `errorMessage` is not related to the title:
https://github.com/qmonmert/generator-jhipster/blob/a854f9a13767b510dfc58719ec3527259fed3977/generators/client/templates/angular/src/main/webapp/app/layouts/error/error.component.ts.ejs#L46-L53

```suggestion
    data: {
      errorMessage: 'error.http.403',
    },
    title: 'error.title',
```

You can use the specific title if you want.

File: generators/client/templates/angular/src/main/webapp/app/layouts/error/error.route.ts.ejs [hunk 4]
```diff
@@ -24,25 +24,17 @@ export const errorRoute: Routes = [
...
   },
   {
     path: '404',
     component: ErrorComponent,
-    data: {
-      pageTitle: 'error.title',
-      errorMessage: 'error.http.404',
-    },
+    title: 'error.http.404',
```
Comment (mshima) [inline generators/client/templates/angular/src/main/webapp/app/layouts/error/error.route.ts.ejs:37]: Same as above

**Comment by mshima** [comment]
Missing `login.route.ts`
Captura de Tela 2022-09-19 às 19 44 05

Translations are ok now:
Captura de Tela 2022-09-19 às 19 45 21

**Comment by qmonmert** [comment]
@mshima I update PR but not sure at all of the last commit

**Comment by mshima** [comment]
I've updated to test.

**Comment by qmonmert** [comment]
> I've updated to test.

thanks

**Comment by mshima** [comment]
@qmonmert do you think the `app` dir is the best place for this component? `app/layout`?

**Review by mshima — APPROVED** [review]
Looks good. Maybe putting the title strategy at `app/layout` and removing the template from the name would make sense.

**Comment by qmonmert** [comment]
> @qmonmert do you think the `app` dir is the best place for this component? `app/layout` maybe?

I thought about that when I coded and I was not sure :/

**Comment by qmonmert** [comment]
maybe we can change that in an other PR when we will be sure

**Comment by mshima** [comment]
@qmonmert please rename from TemplatePageTitleStrategy to AppPageTitleStrategy.
It will keep others files pattern.

**Comment by gmarziou** [comment]
@qmonmert Do you know how this property renaming from `pageTitle` to `title` will be handled by `jhipster upgrade`?

**Pull request merged**